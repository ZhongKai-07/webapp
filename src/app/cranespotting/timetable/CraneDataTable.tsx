import { Feature, FeatureCollection, Point } from 'geojson';

import ToggleSlider from '@/lib/ToggleSlider';

interface CraneTableRowProps {
  crane: Feature<Point>;
  isSelected: boolean;
  // onSelect: (timestamp: number) => void;
  onSelect: () => void;
  count: number;
}

interface CraneDataTableProps {
  cranesFilteredByDay?: FeatureCollection<Point>;
  allCranes?: FeatureCollection<Point>;
  // selectedTimestamp: number;
  selectedTimestamps: number[];
  // onSelect: (timestamp: number) => void;
  onSelect: (timestamp: number[]) => void;
  testId?: string;
}

const CraneTableRow = ({
  crane,
  isSelected,
  onSelect,
  count,
}: CraneTableRowProps) => (
  <tr className='border-b-2 border-slate-750 rounded'>
    <td className='whitespace-nowrap px-6 py-4 text-sm font-medium text-gray-900'>
      <ToggleSlider checked={isSelected} onClick={onSelect} />
    </td>
    <td className='whitespace-nowrap px-6 py-4 text-lg font-medium text-gray-900'>
      {new Date(crane.properties?.timestamp * 1000).toLocaleTimeString()}
    </td>
    <td className='whitespace-nowrap px-6 py-4 text-lg text-gray-900'>
      {count} crane{count > 1 ? 's' : ''}
    </td>
  </tr>
);

// <input
//        type="radio"
//        name="selectedTime"
//        checked={isSelected}
//        onChange={onSelect}
//      />

export const CraneDataTable: React.FC<CraneDataTableProps> = ({
  cranesFilteredByDay,
  // selectedTimestamp,
  selectedTimestamps,
  onSelect,
  allCranes,
  testId,
}) => {
  return (
    <table style={{ width: '100%' }} data-testid={testId}>
      <thead className='bg-gray-100'>
        <tr>
          <th className='px-6 py-3 text-left text-xs font-medium uppercase tracking-wider text-gray-500'>
            Select
          </th>
          <th className='px-6 py-3 text-left text-xs font-medium uppercase tracking-wider text-gray-500'>
            Time
          </th>
          <th className='px-6 py-3 text-left text-xs font-medium uppercase tracking-wider text-gray-500'>
            Cranes detected
          </th>
        </tr>
      </thead>
      <tbody className='bg-white'>
        {cranesFilteredByDay?.features?.map((crane, index) => (
          <CraneTableRow
            key={index}
            crane={crane}
            isSelected={selectedTimestamps.includes(
              crane.properties?.timestamp
            )}
            // isSelected={crane.properties?.timestamp === selectedTimestamp}
            onSelect={() =>
              onSelect(
                selectedTimestamps.includes(crane.properties?.timestamp)
                  ? selectedTimestamps.filter(
                      (timestamp) => timestamp != crane.properties?.timestamp
                    )
                  : [...selectedTimestamps, crane.properties?.timestamp]
              )
            }
            // onSelect
            // }
            count={
              allCranes?.features.filter(
                (feature) =>
                  feature?.properties?.timestamp === crane.properties?.timestamp
              ).length || 0
            }
          />
        ))}
      </tbody>
    </table>
  );
};
