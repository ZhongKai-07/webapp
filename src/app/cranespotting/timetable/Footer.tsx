export const Footer: React.FC<{ lastRefreshTime: string; testId?: string }> = ({
  lastRefreshTime,
  testId,
}) => (
  <div
    className='bg-white p-2 text-center text-sm text-gray-600'
    data-testid={testId}
  >
    Last Data refresh: {lastRefreshTime}
  </div>
);
