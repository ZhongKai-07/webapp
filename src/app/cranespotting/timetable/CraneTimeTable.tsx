import { Feature, FeatureCollection, Point } from 'geojson';
import React, { useEffect, useState } from 'react';

import TableLoading from '@/lib/TableLoading';

import { CraneDataTable } from '@/app/cranespotting/timetable/CraneDataTable';
import { DateSelector } from '@/app/cranespotting/timetable/DateSelector';
import { Footer } from '@/app/cranespotting/timetable/Footer';

interface CraneTimeTableProps {
  cranes?: FeatureCollection<Point>;
  craneTimestamps: number[];
  setCraneTimestamps: React.Dispatch<React.SetStateAction<number[]>>;
  testId?: string;
}

export const isOnSameDay = (
  date1Num: number | undefined,
  date2Num: number | undefined
) => {
  const date1 = date1Num ? new Date(date1Num * 1000) : null;
  const date2 = date2Num ? new Date(date2Num * 1000) : null;
  if (!date1 || !date2) return false;
  return (
    date1.getDate() === date2.getDate() &&
    date1.getMonth() === date2.getMonth() &&
    date1.getFullYear() === date2.getFullYear()
  );
};

export const calculateTimestampsByDay = (timestamps: number[]) =>
  timestamps.filter((timestamp, index) => {
    const currentDate = new Date(timestamp * 1000);
    const previousDate =
      index > 0 ? new Date(timestamps[index - 1] * 1000) : null;
    return (
      previousDate === null ||
      currentDate.getDate() !== previousDate.getDate() ||
      currentDate.getMonth() !== previousDate.getMonth() ||
      currentDate.getFullYear() !== previousDate.getFullYear()
    );
  });

const CraneTimeTable: React.FC<CraneTimeTableProps> = ({
  cranes,
  craneTimestamps,
  setCraneTimestamps,
  testId,
}) => {
  const [deduplicatedCranes, setDeduplicatedCranes] =
    useState<FeatureCollection<Point>>();
  const [displayedCranes, setDisplayedCranes] =
    useState<FeatureCollection<Point>>();
  const [craneDates, setCraneDates] = useState<number[]>([]);
  const [craneSelectedDay, setCraneSelectedDay] = useState<number>();

  useEffect(() => {
    if (craneSelectedDay && deduplicatedCranes) {
      setDisplayedCranes({
        type: 'FeatureCollection',
        features: deduplicatedCranes.features.filter((feature) =>
          isOnSameDay(feature.properties?.timestamp, craneSelectedDay)
        ),
      });
    }
  }, [craneSelectedDay, deduplicatedCranes]);

  useEffect(() => {
    const timestamps = Array.from(
      new Set(
        (cranes as FeatureCollection<Point>)?.features
          .map((feature) => feature.properties?.timestamp)
          .filter((el) => !!el)
      )
    ) as number[];
    const deduplicatedCranes: FeatureCollection<Point> = {
      type: 'FeatureCollection',
      features: timestamps
        .map((timestamp) =>
          cranes?.features.find(
            (feature) => feature.properties?.timestamp === timestamp
          )
        )
        .filter(Boolean) as Feature<Point>[],
    };
    setDeduplicatedCranes(deduplicatedCranes);
    setDisplayedCranes(deduplicatedCranes);

    const timestampsByDay = calculateTimestampsByDay(timestamps);
    setCraneDates(timestampsByDay);
    setCraneSelectedDay(timestampsByDay[0]);
  }, [cranes]);

  if (!cranes) return <TableLoading />;
  return (
    <div className='flex flex-col' data-testid={testId}>
      <DateSelector
        dates={craneDates}
        selectedDay={craneSelectedDay}
        onSelect={setCraneSelectedDay}
        testId='date-selector'
      />
      <CraneDataTable
        cranesFilteredByDay={displayedCranes}
        allCranes={cranes}
        selectedTimestamps={craneTimestamps}
        onSelect={setCraneTimestamps}
        testId='crane-data-table'
      />
      <Footer
        lastRefreshTime={new Date(craneTimestamps[0] * 1000).toLocaleString()}
        testId='footer'
      />
      <div className='flex w-full justify-center'>
        <button
          className='rounded-full w-1/3 content-center  px-5 py-1 text-base font-medium text-gray-700 border-2 border-apian-yellow active:bg-apian-yellow hover:bg-apian-yellow-opacity-50'
          onClick={() => {
            setCraneTimestamps([]);
          }}
        >
          Clear Selections - (
          {
            cranes?.features?.filter((feature) =>
              craneTimestamps.includes(feature.properties?.timestamp)
            ).length
          }{' '}
          selected)
        </button>
      </div>
    </div>
  );
};

export default CraneTimeTable;
