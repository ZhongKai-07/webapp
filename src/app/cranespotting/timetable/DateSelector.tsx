import { useState } from 'react';

import { isOnSameDay } from '@/app/cranespotting/timetable/CraneTimeTable';

interface DateButtonProps {
  date: number;
  isSelected: boolean;
  onSelect: (timestamp: number) => void;
}

interface DateSelectorProps {
  dates: number[];
  selectedDay?: number;
  onSelect: (timestamp: number) => void;
  testId?: string;
}
const DateButton = ({ date, isSelected, onSelect }: DateButtonProps) => {
  const [pressed, setPressed] = useState<boolean>(false);
  return (
    <button
      className={`rounded-full px-5 py-1 text-base font-medium text-gray-700 border-2 border-apian-yellow ${
        isSelected
          ? 'bg-apian-yellow'
          : 'bg-white hover:bg-apian-yellow-opacity-50'
      } ${pressed ? 'shadow-inner shadow' : 'shadow'}`}
      key={date}
      onMouseDown={() => setPressed(true)}
      onMouseUp={() => setPressed(false)}
      onClick={() => onSelect(date)}
    >
      {new Date(date * 1000).toLocaleDateString()}
    </button>
  );
};

export const DateSelector: React.FC<DateSelectorProps> = ({
  dates,
  selectedDay,
  onSelect,
  testId,
}) => (
  <div className='flex justify-center space-x-2 p-3' data-testid={testId}>
    {dates.map((date) => (
      <DateButton
        key={date}
        date={date}
        isSelected={isOnSameDay(date, selectedDay)}
        onSelect={onSelect}
      />
    ))}
  </div>
);
