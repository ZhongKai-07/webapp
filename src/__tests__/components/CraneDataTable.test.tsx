import { cleanup, render } from '@testing-library/react';
import { FeatureCollection, Point } from 'geojson';
import React from 'react';

import { CraneDataTable } from '@/app/cranespotting/timetable/CraneDataTable';

const cranesFilteredByDay: FeatureCollection<Point> = {
  type: 'FeatureCollection',
  features: [
    {
      id: '0',
      type: 'Feature',
      properties: {
        craneId: '1',
        height: 187,
        timestamp: 1705190400,
      },
      geometry: {
        type: 'Point',
        coordinates: [-0.11433123536651221, 51.497263819259494],
      },
    },
    {
      id: '1',
      type: 'Feature',
      properties: {
        craneId: '6',
        height: 61,
        timestamp: 1705276801,
      },
      geometry: {
        type: 'Point',
        coordinates: [-0.10701641144000305, 51.50180701131141],
      },
    },
    {
      id: '2',
      type: 'Feature',
      properties: {
        craneId: '21',
        height: 147,
        timestamp: 1705363201,
      },
      geometry: {
        type: 'Point',
        coordinates: [-0.1174855545800235, 51.5008315750743],
      },
    },
  ],
};

describe('CraneDataTable', () => {
  afterEach(cleanup);

  test('renders CraneDataTable with entire dataset correctly', () => {
    const selectedTimestamp = 1629878400;

    const { container } = render(
      <CraneDataTable
        cranesFilteredByDay={cranesFilteredByDay}
        selectedTimestamps={[selectedTimestamp]}
        // eslint-disable-next-line @typescript-eslint/no-empty-function
        onSelect={() => {}}
        allCranes={cranesFilteredByDay}
      />
    );

    expect(container).toMatchSnapshot();
  });

  test('renders no cranes when cranesFilteredByDay is empty', () => {
    const { container } = render(
      <CraneDataTable
        cranesFilteredByDay={{ type: 'FeatureCollection', features: [] }}
        selectedTimestamps={[]}
        // eslint-disable-next-line @typescript-eslint/no-empty-function
        onSelect={() => {}}
        allCranes={cranesFilteredByDay}
      />
    );

    expect(container).toMatchSnapshot();
    // Account for header
    expect(container.querySelectorAll('tr')).toHaveLength(1);
  });

  // test('calls onSelect with correct timestamp when a row is selected', () => {
  //   const onSelect = jest.fn();

  //   const { getByText } = render(
  //     <CraneDataTable
  //       cranesFilteredByDay={cranesFilteredByDay}
  //       selectedTimestamps={[]}
  //       onSelect={onSelect}
  //       allCranes={cranesFilteredByDay}
  //     />
  //   );

  //   // console.log(container);

  //   const firstRowTimeCell = getByText('12:00:00 AM'); // Adjust this text to match how your component displays the timestamp

  //   const tr = firstRowTimeCell.closest('tr');
  //   if (!tr) {
  //     throw new Error('tr not found');
  //   }
  //   fireEvent.click(tr);
  //   console.log(tr.innerHTML);

  //   // The expected timestamp should match the first object in your cranesFilteredByDay array
  //   expect(onSelect).toHaveBeenCalledWith(
  //     cranesFilteredByDay.features[0].properties?.timestamp
  //   );
  // });

  test('renders individual days correctly', () => {
    const testWithIndex = (
      ts: number,
      filteredCranes: FeatureCollection<Point>
    ) => {
      const { container } = render(
        <CraneDataTable
          cranesFilteredByDay={filteredCranes}
          selectedTimestamps={[ts]}
          // eslint-disable-next-line @typescript-eslint/no-empty-function
          onSelect={() => {}}
          allCranes={cranesFilteredByDay}
        />
      );
      expect(container).toMatchSnapshot();
    };

    cranesFilteredByDay.features.forEach((feature, index) => {
      const selectedTimestamp =
        cranesFilteredByDay.features[index]?.properties?.timestamp;
      const filteredCranes: FeatureCollection<Point> = {
        type: 'FeatureCollection',
        features: cranesFilteredByDay.features.filter(
          (el) => el.properties?.timestamp === selectedTimestamp
        ),
      };

      testWithIndex(selectedTimestamp, filteredCranes);
    });
  });
});
