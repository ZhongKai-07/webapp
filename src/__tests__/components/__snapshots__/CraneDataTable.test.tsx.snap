// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CraneDataTable renders CraneDataTable with entire dataset correctly 1`] = `
<div>
  <table
    style="width: 100%;"
  >
    <thead
      class="bg-gray-100"
    >
      <tr>
        <th
          class="px-6 py-3 text-left text-xs font-medium uppercase tracking-wider text-gray-500"
        >
          Select
        </th>
        <th
          class="px-6 py-3 text-left text-xs font-medium uppercase tracking-wider text-gray-500"
        >
          Time
        </th>
        <th
          class="px-6 py-3 text-left text-xs font-medium uppercase tracking-wider text-gray-500"
        >
          Cranes detected
        </th>
      </tr>
    </thead>
    <tbody
      class="bg-white"
    >
      <tr
        class="border-b-2 border-slate-750 rounded"
      >
        <td
          class="whitespace-nowrap px-6 py-4 text-sm font-medium text-gray-900"
        >
          <div
            class="grid items-center w-[5em] grid-rows-1 grid-cols-1"
          >
            <div
              class="w-[3em] h-[10px] border-1 rounded-[1em] shadow-inner bg-slate-200"
              style="grid-row: 1; grid-column: 1;"
            />
            <div
              class="bg-white shadow  border-[1px] border-slate-300: float:top h-[20px] w-[20px] rounded-full"
              style="grid-row: 1; grid-column: 1; margin-left: -10px; transition-delay: 500ms; transition: margin-left 500ms ease;"
            />
          </div>
        </td>
        <td
          class="whitespace-nowrap px-6 py-4 text-lg font-medium text-gray-900"
        >
          12:00:00 AM
        </td>
        <td
          class="whitespace-nowrap px-6 py-4 text-lg text-gray-900"
        >
          1
           crane
        </td>
      </tr>
      <tr
        class="border-b-2 border-slate-750 rounded"
      >
        <td
          class="whitespace-nowrap px-6 py-4 text-sm font-medium text-gray-900"
        >
          <div
            class="grid items-center w-[5em] grid-rows-1 grid-cols-1"
          >
            <div
              class="w-[3em] h-[10px] border-1 rounded-[1em] shadow-inner bg-slate-200"
              style="grid-row: 1; grid-column: 1;"
            />
            <div
              class="bg-white shadow  border-[1px] border-slate-300: float:top h-[20px] w-[20px] rounded-full"
              style="grid-row: 1; grid-column: 1; margin-left: -10px; transition-delay: 500ms; transition: margin-left 500ms ease;"
            />
          </div>
        </td>
        <td
          class="whitespace-nowrap px-6 py-4 text-lg font-medium text-gray-900"
        >
          12:00:01 AM
        </td>
        <td
          class="whitespace-nowrap px-6 py-4 text-lg text-gray-900"
        >
          1
           crane
        </td>
      </tr>
      <tr
        class="border-b-2 border-slate-750 rounded"
      >
        <td
          class="whitespace-nowrap px-6 py-4 text-sm font-medium text-gray-900"
        >
          <div
            class="grid items-center w-[5em] grid-rows-1 grid-cols-1"
          >
            <div
              class="w-[3em] h-[10px] border-1 rounded-[1em] shadow-inner bg-slate-200"
              style="grid-row: 1; grid-column: 1;"
            />
            <div
              class="bg-white shadow  border-[1px] border-slate-300: float:top h-[20px] w-[20px] rounded-full"
              style="grid-row: 1; grid-column: 1; margin-left: -10px; transition-delay: 500ms; transition: margin-left 500ms ease;"
            />
          </div>
        </td>
        <td
          class="whitespace-nowrap px-6 py-4 text-lg font-medium text-gray-900"
        >
          12:00:01 AM
        </td>
        <td
          class="whitespace-nowrap px-6 py-4 text-lg text-gray-900"
        >
          1
           crane
        </td>
      </tr>
    </tbody>
  </table>
</div>
`;

exports[`CraneDataTable renders individual days correctly 1`] = `
<div>
  <table
    style="width: 100%;"
  >
    <thead
      class="bg-gray-100"
    >
      <tr>
        <th
          class="px-6 py-3 text-left text-xs font-medium uppercase tracking-wider text-gray-500"
        >
          Select
        </th>
        <th
          class="px-6 py-3 text-left text-xs font-medium uppercase tracking-wider text-gray-500"
        >
          Time
        </th>
        <th
          class="px-6 py-3 text-left text-xs font-medium uppercase tracking-wider text-gray-500"
        >
          Cranes detected
        </th>
      </tr>
    </thead>
    <tbody
      class="bg-white"
    >
      <tr
        class="border-b-2 border-slate-750 rounded"
      >
        <td
          class="whitespace-nowrap px-6 py-4 text-sm font-medium text-gray-900"
        >
          <div
            class="grid items-center w-[5em] grid-rows-1 grid-cols-1"
          >
            <div
              class="w-[3em] h-[10px] border-1 rounded-[1em] shadow-inner bg-apian-yellow"
              style="grid-row: 1; grid-column: 1;"
            />
            <div
              class="bg-white shadow  border-[1px] border-slate-300: float:top h-[20px] w-[20px] rounded-full"
              style="grid-row: 1; grid-column: 1; transition-delay: 500ms; transition: margin-left 500ms ease;"
            />
          </div>
        </td>
        <td
          class="whitespace-nowrap px-6 py-4 text-lg font-medium text-gray-900"
        >
          12:00:00 AM
        </td>
        <td
          class="whitespace-nowrap px-6 py-4 text-lg text-gray-900"
        >
          1
           crane
        </td>
      </tr>
    </tbody>
  </table>
</div>
`;

exports[`CraneDataTable renders individual days correctly 2`] = `
<div>
  <table
    style="width: 100%;"
  >
    <thead
      class="bg-gray-100"
    >
      <tr>
        <th
          class="px-6 py-3 text-left text-xs font-medium uppercase tracking-wider text-gray-500"
        >
          Select
        </th>
        <th
          class="px-6 py-3 text-left text-xs font-medium uppercase tracking-wider text-gray-500"
        >
          Time
        </th>
        <th
          class="px-6 py-3 text-left text-xs font-medium uppercase tracking-wider text-gray-500"
        >
          Cranes detected
        </th>
      </tr>
    </thead>
    <tbody
      class="bg-white"
    >
      <tr
        class="border-b-2 border-slate-750 rounded"
      >
        <td
          class="whitespace-nowrap px-6 py-4 text-sm font-medium text-gray-900"
        >
          <div
            class="grid items-center w-[5em] grid-rows-1 grid-cols-1"
          >
            <div
              class="w-[3em] h-[10px] border-1 rounded-[1em] shadow-inner bg-apian-yellow"
              style="grid-row: 1; grid-column: 1;"
            />
            <div
              class="bg-white shadow  border-[1px] border-slate-300: float:top h-[20px] w-[20px] rounded-full"
              style="grid-row: 1; grid-column: 1; transition-delay: 500ms; transition: margin-left 500ms ease;"
            />
          </div>
        </td>
        <td
          class="whitespace-nowrap px-6 py-4 text-lg font-medium text-gray-900"
        >
          12:00:01 AM
        </td>
        <td
          class="whitespace-nowrap px-6 py-4 text-lg text-gray-900"
        >
          1
           crane
        </td>
      </tr>
    </tbody>
  </table>
</div>
`;

exports[`CraneDataTable renders individual days correctly 3`] = `
<div>
  <table
    style="width: 100%;"
  >
    <thead
      class="bg-gray-100"
    >
      <tr>
        <th
          class="px-6 py-3 text-left text-xs font-medium uppercase tracking-wider text-gray-500"
        >
          Select
        </th>
        <th
          class="px-6 py-3 text-left text-xs font-medium uppercase tracking-wider text-gray-500"
        >
          Time
        </th>
        <th
          class="px-6 py-3 text-left text-xs font-medium uppercase tracking-wider text-gray-500"
        >
          Cranes detected
        </th>
      </tr>
    </thead>
    <tbody
      class="bg-white"
    >
      <tr
        class="border-b-2 border-slate-750 rounded"
      >
        <td
          class="whitespace-nowrap px-6 py-4 text-sm font-medium text-gray-900"
        >
          <div
            class="grid items-center w-[5em] grid-rows-1 grid-cols-1"
          >
            <div
              class="w-[3em] h-[10px] border-1 rounded-[1em] shadow-inner bg-apian-yellow"
              style="grid-row: 1; grid-column: 1;"
            />
            <div
              class="bg-white shadow  border-[1px] border-slate-300: float:top h-[20px] w-[20px] rounded-full"
              style="grid-row: 1; grid-column: 1; transition-delay: 500ms; transition: margin-left 500ms ease;"
            />
          </div>
        </td>
        <td
          class="whitespace-nowrap px-6 py-4 text-lg font-medium text-gray-900"
        >
          12:00:01 AM
        </td>
        <td
          class="whitespace-nowrap px-6 py-4 text-lg text-gray-900"
        >
          1
           crane
        </td>
      </tr>
    </tbody>
  </table>
</div>
`;

exports[`CraneDataTable renders no cranes when cranesFilteredByDay is empty 1`] = `
<div>
  <table
    style="width: 100%;"
  >
    <thead
      class="bg-gray-100"
    >
      <tr>
        <th
          class="px-6 py-3 text-left text-xs font-medium uppercase tracking-wider text-gray-500"
        >
          Select
        </th>
        <th
          class="px-6 py-3 text-left text-xs font-medium uppercase tracking-wider text-gray-500"
        >
          Time
        </th>
        <th
          class="px-6 py-3 text-left text-xs font-medium uppercase tracking-wider text-gray-500"
        >
          Cranes detected
        </th>
      </tr>
    </thead>
    <tbody
      class="bg-white"
    />
  </table>
</div>
`;
